services/s3.cfc

```diff
@@ -290,6 +290,172 @@ component {
+    }
+
+    public any function putPart(
+        required string Bucket,
+        required string Key,
+        required string UploadId,
+        required numeric PartNumber, 
+        required any body
+
+    ){
+        var requestSettings = api.resolveRequestSettings( argumentCollection = arguments );
+        dump("putPart");
+        var queryParams = { 
+            'partNumber': arguments.partNumber, 
+            "uploadId": arguments.uploadId 
+        };
+
+        for (
+            var key in arguments
+        ) {
+            // if ( len( arguments[ key ] ) ) queryParams[ utils.parseKey( key ) ] = arguments[ key ];
+        }
+        var apiResponse = apiCall(
+            requestSettings,
+            'PUT',
+            '/#arguments.Key#',
+            queryParams,
+            {
+                // "x-amz-acl": "public-read"
+            },
+            arguments.body
+        );
+
+        if ( apiResponse.statusCode == 200 ) {
+            // apiResponse[ 'data' ] = utils.parseXmlDocument( apiResponse.rawData );
+        }
+        return apiResponse;
+        
+    }
+
+
+    public any function completeMultipartUpload(
+        required string Bucket,
+        required string Key,
+        required string UploadId,
+        required array Parts
+    ){
+
+        var requestSettings = api.resolveRequestSettings( argumentCollection = arguments );
+        var payload = getPartsPayload(Parts);
+        var queryParams = {
+            "uploadId": arguments.UploadId
+        };
+        for (
+            var key in arguments
+        ) {
+            // if(key eq "Parts"){
+            //     continue;
+            // }
+
+            // if ( len( arguments[ key ] ) ) queryParams[ utils.parseKey( key ) ] = arguments[ key ];
+        }
+        var apiResponse = apiCall(
+            requestSettings,
+            'POST',
+            '/#arguments.Key#',
+            queryParams,
+            {},
+            toString(payload)
+           
+        );
+
+
+        if ( apiResponse.statusCode == 200 ) {
+            // apiResponse[ 'data' ] = utils.parseXmlDocument( apiResponse.rawData );
+        }
+        return apiResponse;
+        
+    }
+
+    private xml function getPartsPayload(required Array Parts){
+
+        var ret = '<?xml version="1.0" encoding="UTF-8"?>';
+            ret &= '<CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/">';
+    
+        for (var part in arguments.Parts){
+            ret &= "<Part><ETag>#part.ETag#</ETag><PartNumber>#part.PartNumber#</PartNumber></Part>";
+        }
+
+        ret &= '</CompleteMultipartUpload>'
+        return XMLParse(ret);
+    }
+
+    // ** modified version of generate presigned. Just for putting uploads
+
+    public string function generatePresignedURLForPart(
+        required string Bucket,
+        required string Key,
+        required numeric partNumber,
+        required string uploadId,
+        // required string ContentType,
```

Comment:
Is this needed?

Reply:
REmoved. I thought it was needed as part of the signing but apparently not.